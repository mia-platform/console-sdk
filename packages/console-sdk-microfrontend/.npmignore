.github
.storybook
.yarn
.vscode
*.tgz
**/coverage
.editorconfig
.eslintignore
.eslintrc.cjs
.gitattributes
.gitignore
.nvmrc
.yarnrc.yml
.changeset
CONTRIBUTING.md
index.html
jest.config.js
tsconfig.json
tsconfig.node.json
vite.config.ts
yarn.lock
stats.html
storybook-static
__tests__
node_modules
src/*