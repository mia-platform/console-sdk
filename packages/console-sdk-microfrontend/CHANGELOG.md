# @mia-platform/console-sdk-microfrontend

## 0.0.7

### Patch Changes

- Updated dependencies []:
  - @mia-platform/console-types@0.18.4

## 0.0.6

### Patch Changes

- Updated dependencies []:
  - @mia-platform/console-types@0.18.3

## 0.0.5

### Patch Changes

- Updated dependencies []:
  - @mia-platform/console-types@0.18.2

## 0.0.4

### Patch Changes

- Update dependencies

## 0.0.3

### Patch Changes

- Update dependencies

## 0.0.2

### Added

- ðŸš€ New getMicrofrontendNode to console-sdk

## 0.0.1

### Added

- ðŸš€ Setup repo
