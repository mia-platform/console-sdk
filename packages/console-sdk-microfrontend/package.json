{
  "name": "@mia-platform/console-sdk-microfrontend",
  "license": "Apache-2.0",
  "version": "0.0.9",
  "main": "./build/index.js",
  "typings": "./build/index.d.ts",
  "module": "./build/index.js",
  "scripts": {
    "preinstall": "npx only-allow pnpm",
    "build": "rm -rf ./build && tsc --build",
    "lint": "eslint . --ext .ts",
    "test": "jest --coverage",
    "unit": "jest"
  },
  "dependencies": {
    "@mia-platform/console-types": "workspace:^",
    "rxjs": "^7.8.1",
    "ramda": "^0.29.1"
  },
  "devDependencies": {
    "@typescript-eslint/eslint-plugin": "^7.2.0",
    "@typescript-eslint/parser": "^7.1.0",
    "@types/ramda": "^0.29.11",
    "typescript": "^5.3.3",
    "vite-plugin-qiankun": "1.0.15",
    "@types/jest": "^29.5.12",
    "@types/jsdom": "21.1.6",
    "jest": "^29.7.0",
    "ts-jest": "^29.1.2"
  },
  "engines": {
    "node": ">=v18.18.2"
  },
  "homepage": "https://github.com/mia-platform/console-sdk-microfrontend#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/mia-platform/console-sdk-microfrontend.git"
  },
  "bugs": {
    "url": "https://github.com/mia-platform/console-sdk-microfrontend/issues"
  },
  "jest": {
    "preset": "ts-jest",
    "modulePathIgnorePatterns": [
      "<rootDir>/build"
    ],
    "testMatch": [
      "**/__tests__/**/*.+(ts)",
      "**/?(*.)+(spec|test).+(ts)"
    ]
  }
}
