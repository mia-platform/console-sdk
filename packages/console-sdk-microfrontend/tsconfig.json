{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "./build",
    "rootDir": "src/"
  },
  "exclude": [
    "./node_modules",
    "./coverage",
    "./build",
    "./**/*.test.ts",
    "tsconfig.json"
  ],
  "include": [
    "./**/*.ts",
    "./**/*.json"
  ],
}
