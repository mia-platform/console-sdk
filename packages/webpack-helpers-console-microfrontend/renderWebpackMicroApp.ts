
export default function renderWebpackMicroApp (
  mount: () => void,
  unmount: () => void,
  bootstrap: () => void
) {
  return {
    mount,
    unmount,
    bootstrap
  }
}