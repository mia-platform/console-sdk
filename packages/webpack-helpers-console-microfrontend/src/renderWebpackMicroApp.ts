/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { ConsoleSDK, IConsoleSDK } from '@mia-platform/microfrontend-sdk'
import { QiankunLifeCycle, QiankunProps, qiankunWindow } from 'vite-plugin-qiankun/dist/helper'

export type ConsoleLifecycleFunction = (
  isConnectedToConsole: boolean,
  consoleSDK: IConsoleSDK,
) => void

export type IViteParams = {
  mount: ConsoleLifecycleFunction
  unmount: ConsoleLifecycleFunction
  update?: ConsoleLifecycleFunction
  bootstrap?: () => void
}

export function getSDK(props: QiankunProps): {
  consoleSDK: IConsoleSDK
  isConnectedToConsole: boolean
} {
  const consoleSDK = new ConsoleSDK(props)
  const isConnectedToConsole = Boolean(qiankunWindow.__POWERED_BY_QIANKUN__)

  return {
    consoleSDK,
    isConnectedToConsole,
  }
}

export const decorateLifecycleFunction = (lifecycleFunction: IViteParams[keyof IViteParams]) => {
  return (props: QiankunProps): void => {
    const { isConnectedToConsole, consoleSDK } = getSDK(props)
    if (lifecycleFunction) {
      lifecycleFunction(isConnectedToConsole, consoleSDK)
    }
  }
}

export default function renderWebpackMicroApp({ mount, unmount, bootstrap, update }: IViteParams): QiankunLifeCycle {
  return {
    mount: decorateLifecycleFunction(mount),
    unmount: decorateLifecycleFunction(unmount),
    update: decorateLifecycleFunction(update),
    ...bootstrap ? { bootstrap } : { bootstrap: () => ({}) },
  }
}
