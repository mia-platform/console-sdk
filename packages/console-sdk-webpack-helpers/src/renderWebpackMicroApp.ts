/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { ConsoleSDK, IConsoleProps, IMicrofrontendIntegrator } from '@mia-platform/console-sdk-microfrontend'
import { QiankunLifeCycle, QiankunProps, qiankunWindow } from 'vite-plugin-qiankun/dist/helper'

export type ConsoleLifecycleFunction = (
  isConnectedToConsole: boolean,
  consoleSDK: IMicrofrontendIntegrator | undefined,
) => void

export type IViteParams = {
  mount: ConsoleLifecycleFunction
  unmount: ConsoleLifecycleFunction
  update?: ConsoleLifecycleFunction
  bootstrap?: () => void
}

type SDKParams = {
  consoleSDK: IMicrofrontendIntegrator | undefined
  isConnectedToConsole: boolean
}

export function getSDK(props: IConsoleProps): SDKParams {
  const consoleSDK = new ConsoleSDK(props)
  const isConnectedToConsole = Boolean(qiankunWindow.__POWERED_BY_QIANKUN__)

  return {
    consoleSDK,
    isConnectedToConsole,
  }
}

export const decorateLifecycleFunction = (lifecycleFunction: IViteParams[keyof IViteParams]) => {
  return (props?: QiankunProps): void => {
    const defaultValues: SDKParams = { isConnectedToConsole: false, consoleSDK: undefined }
    const { isConnectedToConsole, consoleSDK } = props && Object.keys(props).length > 0
      ? getSDK(props as IConsoleProps)
      : defaultValues

    if (lifecycleFunction) {
      lifecycleFunction(isConnectedToConsole, consoleSDK)
    }
  }
}

export default function renderWebpackMicroApp({ mount, unmount, bootstrap, update }: IViteParams): QiankunLifeCycle {
  return {
    mount: decorateLifecycleFunction(mount),
    unmount: decorateLifecycleFunction(unmount),
    update: decorateLifecycleFunction(update),
    ...bootstrap ? { bootstrap } : { bootstrap: () => ({}) },
  }
}
