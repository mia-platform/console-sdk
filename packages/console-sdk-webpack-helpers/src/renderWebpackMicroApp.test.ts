/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { ConsoleSDK, IConsoleProps } from '@mia-platform/console-sdk-microfrontend'
import { JSDOM } from 'jsdom'

import renderWebpackMicroApp, { IViteParams, decorateLifecycleFunction, getSDK } from './renderWebpackMicroApp'

const dom = new JSDOM()
const { document } = dom.window

const quiankunContainer = document.createElement('div')
quiankunContainer.id = '__qiankun_microapp_wrapper_for_microfrontend__'

const consoleProps: IConsoleProps = {
  name: 'testMicrofrontend',
  container: quiankunContainer,
  // some console injected props
  eventListener: jest.fn(),
  featureTogglesProxyContext: {},
  hotkeysContext: {},
  resourceAPI: {
    endpoints: {},
    collections: {},
    configMaps: {},
    services: {},
    unsecretedVariables: [],

    forceConfigUpdateChecksum: '',
    microfrontendPluginConfig: {},

    selectedEnvironment: {},
    selectedProject: {},

    writeConfig: jest.fn(),

    _version: '0.0.0',
    _signals: { mount: jest.fn() },
  },
}

const viteParamsMock: IViteParams = {
  mount: jest.fn(),
  unmount: jest.fn(),
  bootstrap: jest.fn(),
}

describe('Webpack Micro App Rendering', () => {
  it('should create a ConsoleSDK instance', () => {
    const { consoleSDK, isConnectedToConsole } = getSDK(consoleProps)
    expect(consoleSDK).toBeInstanceOf(ConsoleSDK)
    expect(isConnectedToConsole).toBe(false)
  })

  it('should decorate a lifecycle function', () => {
    const decoratedMount = decorateLifecycleFunction(viteParamsMock.mount)
    const decoratedUnmount = decorateLifecycleFunction(viteParamsMock.unmount)
    const decoratedBootstrap = decorateLifecycleFunction(viteParamsMock.bootstrap)

    decoratedMount(consoleProps)
    decoratedUnmount(consoleProps)
    decoratedBootstrap(consoleProps)

    expect(viteParamsMock.mount).toHaveBeenCalledWith(false, expect.any(ConsoleSDK))
    expect(viteParamsMock.unmount).toHaveBeenCalledWith(false, expect.any(ConsoleSDK))
    expect(viteParamsMock.bootstrap).toHaveBeenCalledWith(false, expect.any(ConsoleSDK))
  })

  it('should render the Webpack micro app', () => {
    const { mount, unmount, bootstrap } = renderWebpackMicroApp(viteParamsMock)

    mount(consoleProps)
    unmount(consoleProps)

    if (bootstrap) {
      bootstrap()
    }

    expect(viteParamsMock.mount).toHaveBeenCalledWith(false, expect.any(ConsoleSDK))
    expect(viteParamsMock.unmount).toHaveBeenCalledWith(false, expect.any(ConsoleSDK))
    expect(viteParamsMock.bootstrap).toHaveBeenCalledWith(false, expect.any(ConsoleSDK))
  })
})
