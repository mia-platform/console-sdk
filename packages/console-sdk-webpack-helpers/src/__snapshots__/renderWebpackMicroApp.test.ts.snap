// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webpack Micro App Rendering should mount the micro app on the missing props 1`] = `
[MockFunction] {
  "calls": [
    [
      false,
      ConsoleSDK {
        "configObservable": {
          "_version": "mock_version",
          "collections": Observable {
            "_subscribe": [Function],
          },
          "configMaps": Observable {
            "_subscribe": [Function],
          },
          "endpoints": Observable {
            "_subscribe": [Function],
          },
          "forceConfigUpdateChecksum": Observable {
            "_subscribe": [Function],
          },
          "microfrontendPluginConfig": Observable {
            "_subscribe": [Function],
          },
          "selectedEnvironment": Observable {
            "_subscribe": [Function],
          },
          "selectedProject": Observable {
            "_subscribe": [Function],
          },
          "services": Observable {
            "_subscribe": [Function],
          },
          "unsecretedVariables": Observable {
            "_subscribe": [Function],
          },
        },
        "consoleWriteConfig": [MockFunction],
        "container": <div
          id="__qiankun_microapp_wrapper_for_microfrontend__"
        />,
        "contexts": {
          "featureTogglesProxyContext": {},
          "hotkeysContext": {},
        },
        "events": Subject {
          "closed": false,
          "currentObservers": null,
          "hasError": false,
          "isStopped": false,
          "observers": [
            SafeSubscriber {
              "_finalizers": [
                Subscription {
                  "_finalizers": null,
                  "_parentage": [Circular],
                  "closed": false,
                  "initialTeardown": [Function],
                },
              ],
              "_parentage": null,
              "closed": false,
              "destination": ConsumerObserver {
                "partialObserver": {
                  "complete": undefined,
                  "error": undefined,
                  "next": [MockFunction],
                },
              },
              "initialTeardown": undefined,
              "isStopped": false,
            },
          ],
          "thrownError": null,
        },
        "name": "testMicrofrontend",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Webpack Micro App Rendering should mount the micro app on the missing props 2`] = `
[MockFunction] {
  "calls": [
    [
      false,
      ConsoleSDK {
        "configObservable": {
          "_version": "mock_version",
          "collections": Observable {
            "_subscribe": [Function],
          },
          "configMaps": Observable {
            "_subscribe": [Function],
          },
          "endpoints": Observable {
            "_subscribe": [Function],
          },
          "forceConfigUpdateChecksum": Observable {
            "_subscribe": [Function],
          },
          "microfrontendPluginConfig": Observable {
            "_subscribe": [Function],
          },
          "selectedEnvironment": Observable {
            "_subscribe": [Function],
          },
          "selectedProject": Observable {
            "_subscribe": [Function],
          },
          "services": Observable {
            "_subscribe": [Function],
          },
          "unsecretedVariables": Observable {
            "_subscribe": [Function],
          },
        },
        "consoleWriteConfig": [MockFunction],
        "container": <div
          id="__qiankun_microapp_wrapper_for_microfrontend__"
        />,
        "contexts": {
          "featureTogglesProxyContext": {},
          "hotkeysContext": {},
        },
        "events": Subject {
          "closed": false,
          "currentObservers": null,
          "hasError": false,
          "isStopped": false,
          "observers": [
            SafeSubscriber {
              "_finalizers": [
                Subscription {
                  "_finalizers": null,
                  "_parentage": [Circular],
                  "closed": false,
                  "initialTeardown": [Function],
                },
              ],
              "_parentage": null,
              "closed": false,
              "destination": ConsumerObserver {
                "partialObserver": {
                  "complete": undefined,
                  "error": undefined,
                  "next": [MockFunction],
                },
              },
              "initialTeardown": undefined,
              "isStopped": false,
            },
          ],
          "thrownError": null,
        },
        "name": "testMicrofrontend",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Webpack Micro App Rendering should mount the micro app on the missing props 3`] = `
[MockFunction] {
  "calls": [
    [
      false,
      ConsoleSDK {
        "configObservable": {
          "_version": "mock_version",
          "collections": Observable {
            "_subscribe": [Function],
          },
          "configMaps": Observable {
            "_subscribe": [Function],
          },
          "endpoints": Observable {
            "_subscribe": [Function],
          },
          "forceConfigUpdateChecksum": Observable {
            "_subscribe": [Function],
          },
          "microfrontendPluginConfig": Observable {
            "_subscribe": [Function],
          },
          "selectedEnvironment": Observable {
            "_subscribe": [Function],
          },
          "selectedProject": Observable {
            "_subscribe": [Function],
          },
          "services": Observable {
            "_subscribe": [Function],
          },
          "unsecretedVariables": Observable {
            "_subscribe": [Function],
          },
        },
        "consoleWriteConfig": [MockFunction],
        "container": <div
          id="__qiankun_microapp_wrapper_for_microfrontend__"
        />,
        "contexts": {
          "featureTogglesProxyContext": {},
          "hotkeysContext": {},
        },
        "events": Subject {
          "closed": false,
          "currentObservers": null,
          "hasError": false,
          "isStopped": false,
          "observers": [
            SafeSubscriber {
              "_finalizers": [
                Subscription {
                  "_finalizers": null,
                  "_parentage": [Circular],
                  "closed": false,
                  "initialTeardown": [Function],
                },
              ],
              "_parentage": null,
              "closed": false,
              "destination": ConsumerObserver {
                "partialObserver": {
                  "complete": undefined,
                  "error": undefined,
                  "next": [MockFunction],
                },
              },
              "initialTeardown": undefined,
              "isStopped": false,
            },
          ],
          "thrownError": null,
        },
        "name": "testMicrofrontend",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
